pylops/signalprocessing/_fourierradon2d_cuda.py

```diff
@@ -0,0 +1,86 @@
+from cmath import exp
+from math import pi
+
+from numba import cuda
+
+
+@cuda.jit()
+def _radon_inner_2d_kernel(x, y, f, px, h, flim0, flim1, npx, nh):
+    """Cuda kernels for FourierRadon2D operator
+
+    Cuda implementation of the on-the-fly kernel creation and application for the
+    FourierRadon2D operator. See :class:`pylops.signalprocessing.FourierRadon2D`
+    for details about input parameters.
+
+    """
+    ih, ifr = cuda.grid(2)
+    if ih < nh and ifr >= flim0 and ifr <= flim1:
+        for ipx in range(npx):
+            y[ih, ifr] += x[ipx, ifr] * exp(-1j * 2 * pi * f[ifr] * px[ipx] * h[ih])
```

Comment:
numba is very very annoying with types, making it hard to cast things. I'd love to check if the types here are correct (no float64 when everythign should be float32 for example). I'll try if I have some time. fyi you can do it with: https://numba.readthedocs.io/en/stable/reference/jit-compilation.html#Dispatcher.inspect_types

Reply:
Interesting!

This is what i get after making a change to line 19 into `y[ih, ifr] += x[ipx, ifr] * np.complex64(exp(-1j * 2. * pi * f[ifr] * px[ipx] * h[ih]))` (which I believed was causing the outputs to be complex128 even if inputs were complex64

```
f float32
_ZN6cudapy8__main__22_radon_inner_2d_kernelB2v3B96cw51cXTLSUwv1sCUt9Uw11Ew1dRRQPKiLTj0gIGIFp_2b2oLQFEYYkHSQB1OQAk0Bynm21OizQ1K0UoIGvDpQE8oxrNQE_3dE5ArrayI9complex64Li2E1C7mutable7alignedE5ArrayI9complex64Li2E1C7mutable7alignedE5ArrayIfLi1E1C7mutable7alignedE5ArrayIfLi1E1C7mutable7alignedE5ArrayIfLi1E1C7mutable7alignedExxxx (Array(complex64, 2, 'C', False, aligned=True), Array(complex64, 2, 'C', False, aligned=True), Array(float32, 1, 'C', False, aligned=True), Array(float32, 1, 'C', False, aligned=True), Array(float32, 1, 'C', False, aligned=True), int64, int64, int64, int64)
--------------------------------------------------------------------------------
# File: <ipython-input-3-14e7e7166147>
# --- LINE 8 --- 
# label 0
#   x = arg(0, name=x)  :: array(complex64, 2d, C)
#   y = arg(1, name=y)  :: array(complex64, 2d, C)
#   f = arg(2, name=f)  :: array(float32, 1d, C)
#   px = arg(3, name=px)  :: array(float32, 1d, C)
#   h = arg(4, name=h)  :: array(float32, 1d, C)
#   flim0 = arg(5, name=flim0)  :: int64
#   flim1 = arg(6, name=flim1)  :: int64
#   npx = arg(7, name=npx)  :: int64
#   nh = arg(8, name=nh)  :: int64

@cuda.jit()

# --- LINE 9 --- 

def _radon_inner_2d_kernel(x, y, f, px, h, flim0, flim1, npx, nh):

    # --- LINE 10 --- 

    """Cuda kernels for FourierRadon2D operator

# --- LINE 11 --- 



    # --- LINE 12 --- 

    Cuda implementation of the on-the-fly kernel creation and application for the

    # --- LINE 13 --- 

    FourierRadon2D operator. See :class:`pylops.signalprocessing.FourierRadon2D`

    # --- LINE 14 --- 

    for details about input parameters.

# --- LINE 15 --- 



    # --- LINE 16 --- 

    """

    # --- LINE 17 --- 
    #   $2load_global.0 = global(cuda: <module 'numba.cuda' from '/usr/local/lib/python3.10/dist-packages/numba/cuda/__init__.py'>)  :: Module(<module 'numba.cuda' from '/usr/local/lib/python3.10/dist-packages/numba/cuda/__init__.py'>)
    #   $4load_method.1 = getattr(value=$2load_global.0, attr=grid)  :: Function(<intrinsic grid>)
    #   del $2load_global.0
    #   $const6.2 = const(int, 2)  :: Literal[int](2)
    #   $8call_method.3 = call $4load_method.1($const6.2, func=$4load_method.1, args=[Var($const6.2, <ipython-input-3-14e7e7166147>:17)], kws=(), vararg=None, varkwarg=None, target=None)  :: (int32,) -> UniTuple(int64 x 2)
    #   del $const6.2
    #   del $4load_method.1
    #   $10unpack_sequence.6 = exhaust_iter(value=$8call_method.3, count=2)  :: UniTuple(int64 x 2)
    #   del $8call_method.3
    #   $10unpack_sequence.4 = static_getitem(value=$10unpack_sequence.6, index=0, index_var=None, fn=<built-in function getitem>)  :: int64
    #   $10unpack_sequence.5 = static_getitem(value=$10unpack_sequence.6, index=1, index_var=None, fn=<built-in function getitem>)  :: int64
    #   del $10unpack_sequence.6
    #   ih = $10unpack_sequence.4  :: int64
    #   del $10unpack_sequence.4
    #   ifr = $10unpack_sequence.5  :: int64
    #   del $10unpack_sequence.5

    ih, ifr = cuda.grid(2)

    # --- LINE 18 --- 
    #   $20compare_op.9 = ih < nh  :: bool
    #   del nh
    #   bool22 = global(bool: <class 'bool'>)  :: Function(<class 'bool'>)
    #   $22pred = call bool22($20compare_op.9, func=bool22, args=(Var($20compare_op.9, <ipython-input-3-14e7e7166147>:18),), kws=(), vararg=None, varkwarg=None, target=None)  :: (bool,) -> bool
    #   del bool22
    #   del $20compare_op.9
    #   branch $22pred, 24, 124
    # label 24
    #   del $22pred
    #   $28compare_op.2 = ifr >= flim0  :: bool
    #   del flim0
    #   bool30 = global(bool: <class 'bool'>)  :: Function(<class 'bool'>)
    #   $30pred = call bool30($28compare_op.2, func=bool30, args=(Var($28compare_op.2, <ipython-input-3-14e7e7166147>:18),), kws=(), vararg=None, varkwarg=None, target=None)  :: (bool,) -> bool
    #   del bool30
    #   del $28compare_op.2
    #   branch $30pred, 32, 128
    # label 32
    #   del $30pred
    #   $36compare_op.2 = ifr <= flim1  :: bool
    #   del flim1
    #   bool38 = global(bool: <class 'bool'>)  :: Function(<class 'bool'>)
    #   $38pred = call bool38($36compare_op.2, func=bool38, args=(Var($36compare_op.2, <ipython-input-3-14e7e7166147>:18),), kws=(), vararg=None, varkwarg=None, target=None)  :: (bool,) -> bool
    #   del bool38
    #   del $36compare_op.2
    #   branch $38pred, 40, 132
    # label 124
    #   del y
    #   del x
    #   del px
    #   del npx
    #   del ih
    #   del ifr
    #   del h
    #   del flim1
    #   del flim0
    #   del f
    #   del $22pred
    #   $const124.0 = const(NoneType, None)  :: none
    #   $126return_value.1 = cast(value=$const124.0)  :: none
    #   del $const124.0
    #   return $126return_value.1
    # label 128
    #   del y
    #   del x
    #   del px
    #   del npx
    #   del ih
    #   del ifr
    #   del h
    #   del flim1
    #   del f
    #   del $30pred
    #   $const128.0 = const(NoneType, None)  :: none
    #   $130return_value.1 = cast(value=$const128.0)  :: none
    #   del $const128.0
    #   return $130return_value.1
    # label 132
    #   del y
    #   del x
    #   del px
    #   del npx
    #   del ih
    #   del ifr
    #   del h
    #   del f
    #   del $38pred
    #   $const132.0 = const(NoneType, None)  :: none
    #   $134return_value.1 = cast(value=$const132.0)  :: none
    #   del $const132.0
    #   return $134return_value.1

    if ih < nh and ifr >= flim0 and ifr <= flim1:

        # --- LINE 19 --- 
        # label 40
        #   del $38pred
        #   $40load_global.0 = global(range: <class 'range'>)  :: Function(<class 'range'>)
        #   $44call_function.2 = call $40load_global.0(npx, func=$40load_global.0, args=[Var(npx, <ipython-input-3-14e7e7166147>:8)], kws=(), vararg=None, varkwarg=None, target=None)  :: (int64,) -> range_state_int64
        #   del npx
        #   del $40load_global.0
        #   $46get_iter.3 = getiter(value=$44call_function.2)  :: range_iter_int64
        #   del $44call_function.2
        #   $phi48.0 = $46get_iter.3  :: range_iter_int64
        #   del $46get_iter.3
        #   jump 48
        # label 48
        #   $48for_iter.1 = iternext(value=$phi48.0)  :: pair<int64, bool>
        #   $48for_iter.2 = pair_first(value=$48for_iter.1)  :: int64
        #   $48for_iter.3 = pair_second(value=$48for_iter.1)  :: bool
        #   del $48for_iter.1
        #   $phi50.1 = $48for_iter.2  :: int64
        #   del $48for_iter.2
        #   branch $48for_iter.3, 50, 136
        # label 50
        #   del $48for_iter.3
        #   ipx = $phi50.1  :: int64
        #   del $phi50.1
        # label 136
        #   del y
        #   del x
        #   del px
        #   del ih
        #   del ifr
        #   del h
        #   del f
        #   del $phi50.1
        #   del $phi48.0
        #   del $48for_iter.3
        #   $const136.0 = const(NoneType, None)  :: none
        #   $138return_value.1 = cast(value=$const136.0)  :: none
        #   del $const136.0
        #   return $138return_value.1

        for ipx in range(npx):

            # --- LINE 20 --- 
            #   $58build_tuple.5 = build_tuple(items=[Var(ih, <ipython-input-3-14e7e7166147>:17), Var(ifr, <ipython-input-3-14e7e7166147>:17)])  :: UniTuple(int64 x 2)
            #   $62binary_subscr.8 = getitem(value=y, index=$58build_tuple.5, fn=<built-in function getitem>)  :: complex64
            #   $70build_tuple.12 = build_tuple(items=[Var(ipx, <ipython-input-3-14e7e7166147>:19), Var(ifr, <ipython-input-3-14e7e7166147>:17)])  :: UniTuple(int64 x 2)
            #   $72binary_subscr.13 = getitem(value=x, index=$70build_tuple.12, fn=<built-in function getitem>)  :: complex64
            #   del $70build_tuple.12
            #   $74load_global.14 = global(np: <module 'numpy' from '/usr/local/lib/python3.10/dist-packages/numpy/__init__.py'>)  :: Module(<module 'numpy' from '/usr/local/lib/python3.10/dist-packages/numpy/__init__.py'>)
            #   $76load_method.15 = getattr(value=$74load_global.14, attr=complex64)  :: class(complex64)
            #   del $74load_global.14
            #   $78load_global.16 = global(exp: <built-in function exp>)  :: Function(<built-in function exp>)
            #   $const80.17 = const(complex, -2j)  :: complex128
            #   $82load_global.18 = global(pi: 3.141592653589793)  :: float64
            #   $84binary_multiply.19 = $const80.17 * $82load_global.18  :: complex128
            #   del $const80.17
            #   del $82load_global.18
            #   $90binary_subscr.22 = getitem(value=f, index=ifr, fn=<built-in function getitem>)  :: float32
            #   $92binary_multiply.23 = $84binary_multiply.19 * $90binary_subscr.22  :: complex128
            #   del $90binary_subscr.22
            #   del $84binary_multiply.19
            #   $98binary_subscr.26 = getitem(value=px, index=ipx, fn=<built-in function getitem>)  :: float32
            #   del ipx
            #   $100binary_multiply.27 = $92binary_multiply.23 * $98binary_subscr.26  :: complex128
            #   del $98binary_subscr.26
            #   del $92binary_multiply.23
            #   $106binary_subscr.30 = getitem(value=h, index=ih, fn=<built-in function getitem>)  :: float32
            #   $108binary_multiply.31 = $100binary_multiply.27 * $106binary_subscr.30  :: complex128
            #   del $106binary_subscr.30
            #   del $100binary_multiply.27
            #   $110call_function.32 = call $78load_global.16($108binary_multiply.31, func=$78load_global.16, args=[Var($108binary_multiply.31, <ipython-input-3-14e7e7166147>:20)], kws=(), vararg=None, varkwarg=None, target=None)  :: (complex128,) -> complex128
            #   del $78load_global.16
            #   del $108binary_multiply.31
            #   $112call_method.33 = call $76load_method.15($110call_function.32, func=$76load_method.15, args=[Var($110call_function.32, <ipython-input-3-14e7e7166147>:20)], kws=(), vararg=None, varkwarg=None, target=None)  :: (complex128,) -> complex64
            #   del $76load_method.15
            #   del $110call_function.32
            #   $114binary_multiply.34 = $72binary_subscr.13 * $112call_method.33  :: complex64
            #   del $72binary_subscr.13
            #   del $112call_method.33
            #   $116inplace_add.35 = inplace_binop(fn=<built-in function iadd>, immutable_fn=<built-in function add>, lhs=$62binary_subscr.8, rhs=$114binary_multiply.34, static_lhs=Undefined, static_rhs=Undefined)  :: complex64
            #   del $62binary_subscr.8
            #   del $114binary_multiply.34
            #   y[$58build_tuple.5] = $116inplace_add.35  :: (Array(complex64, 2, 'C', False, aligned=True), UniTuple(int64, 2), complex64) -> none
            #   del $58build_tuple.5
            #   del $116inplace_add.35
            #   jump 48

            y[ih, ifr] += x[ipx, ifr] * np.complex64(exp(-1j * 2. * pi * f[ifr] * px[ipx] * h[ih]))
```

Have a look to see if you agree that now everything remains in complex64, if so I can push this change

Reply:
You still have the complex128 appear because of the float literal. The following should work to eliminate the intermediate cast. It should work for both single (all calculations will be kept in float32/complex64) and double (when complex64 `TWO_PI_I` encounters float64 `f[ifr] * px[ipx] * h[ih]`, it will be promoted to complex128) precision. The `exp` function will be called with a complex64 signature in the first case, and a complex128 signature in the second case.

```python
TWO_PI_I = cp.complex64(-1j * 2. * pi)
@cuda.jit
def ...
    ...
        y[ih, ifr] += x[ipx, ifr] * exp(TWO_PI_I * f[ifr] * px[ipx] * h[ih])
```

If you want to get the most out of performance (we are now entering micro-optimization territory!), you can use CUDA [intrinsics](https://numba.readthedocs.io/en/stable/cuda-reference/libdevice.html) (inspired from [here](https://stackoverflow.com/a/9863048))

```python
TWO_PI = cp.float32(-2. * pi)
I = cp.complex64(1j)

@cuda.jit
def ...
    ...
        s, c = cuda.libdevice.sincosf(TWO_PI * f[ifr] * px[ipx] * h[ih])
        y[ih, ifr] += x[ipx, ifr] * (c + I * s)
```

Reply:
By the way, I don't think `fastmath=True` would trigger the use of sincosf, since it doesn't appear in the [list of optimizations](https://numba.readthedocs.io/en/stable/cuda/fastmath.html#cuda-fast-math). I'd have to confirm this on their forums.

Reply:
Interesting, I didn't catch that!! I'll fix it in the new PR when I introduce the grid strided loops :)